src/utils/view.js

```diff
@@ -0,0 +1,335 @@
+ */
+const negateTransform = transform =>
+  // This ensures +0 instead of -0
+  origin.translate(-transform.x || 0, -transform.y || 0).scale(transform.k);
+
+/**
+ * Sets the extents of the given view.
+ * Translate and scale extents may be provided individually.
+ * @param {Object} view The view
+ * @param {Object} extents The exents
+ */
+export const setViewExtents = (view, { translate, scale }) => {
+  if (translate) {
+    const { minX, minY, maxX, maxY } = translate;
+    view.zoom.translateExtent([[minX, minY], [maxX, maxY]]);
+  }
+
+  if (scale) {
+    const { minK, maxK } = scale;
+    view.zoom.scaleExtent([minK, maxK]);
+  }
+};
+
+/**
+ * Gets the extents of the given view including both translate and scale.
+ * @param {Object} view The view
+ * @returns {Object} The view extents
+ */
+export const getViewExtents = view => {
+  const scale = view.zoom.scaleExtent();
+  const translate = view.zoom.translateExtent();
+  return {
+    translate: {
+      minX: translate[0][0],
+      minY: translate[0][1],
+      maxX: translate[1][0],
+      maxY: translate[1][1]
+    },
+    scale: { minK: scale[0], maxK: scale[1] }
+  };
+};
+
+/**
+ * Gets the bounds of the viewport.
+ * By default this retrieves the container element dimensions unless an alternative viewport is set.
+ * @param {Object} view The view
+ * @returns {Object} The viewport extents
+ */
+export const getViewport = view => {
+  const viewport = view.zoom.extent()(select(view.container.current));
+  return {
+    top: viewport[0][1],
+    left: viewport[0][0],
+    bottom: viewport[1][1],
+    right: viewport[1][0]
+  };
+};
+
+/**
+ * Sets the bounds of the viewport instead of using container element dimensions.
+ * @param {Object} view The view
+ * @param {Object} viewport The viewport bounds
+ */
+export const setViewport = (view, viewport) => {
+  view.zoom.extent([
+    [viewport.left, viewport.top],
+    [viewport.right, viewport.bottom]
+  ]);
+};
+
+/**
+ * Sets the view transform on the view with optional transition.
+ * Always respects the view extent constraints.
+ * Translate and scale components may be provided individually.
+ * Transform is absolute unless relative option set, where it will be added.
+ * @param {Object} view The view
+ * @param {Object} transform The transform to set on the view
+ * @param {?Number} [duration=0] The transition duration (or 0 for none)
+ * @param {?Boolean} [relative=false] When true, transform added to current transform
+ */
+export const setViewTransform = (
+  view,
+  transform,
+  duration = 0,
+  relative = false
+) => {
+  const container = select(view.container.current);
+  const current = getViewTransform(view);
+  const hasTranslation =
+    typeof transform.x !== 'undefined' && typeof transform.y !== 'undefined';
+  let k, x, y;
+
+  if (typeof jest !== 'undefined') {
+    // Transitions not supported in tests
+    duration = 0;
+  }
+
+  if (relative) {
+    // Relative: v = v. + t
```

Comment:
perhaps need to lose the `.` behind the `v` here?

Reply:
I've changed the notation here to make this clearer, so `v'` is used for the view transform after application.